wdl/downsampling_pipeline_part2.wdl

```diff
@@ -1,534 +1,638 @@
+
+  String bam_downsample_name = basename(downsample_file_1, "_downsample_~{final_depth}x.1.fastq") + "_downsample.bam"
+  
+  output {
+    File bam_downsample_file = bam_downsample_name
+  }
+
+  command <<<
+    set -euo pipefail
+
+    #realigns downsampled paired fq files to sam and then bam with bwa
+    NT=$(nproc)
+
+    bwa mem -M \
+      -t ${NT}\
+      ~{reference_fasta} \
+      ~{downsample_file_1} \
+      ~{downsample_file_2} \
+      | samtools view \
+      -bS - > ~{bam_downsample_name}
         
-    >>>
-
-    runtime {
-        docker: downsample_docker
-        memory: mem_size
-        cpu: "5"
-        disks: "local-disk " + disk_size + " HDD"
-    }
+  >>>
+
+  runtime {
+    docker: downsample_docker
+    cpu: select_first([runtime_attr.cpu_cores, default_attr.cpu_cores])
+    memory: select_first([runtime_attr.mem_gb, default_attr.mem_gb]) + " GiB"
+    disks: "local-disk " + select_first([runtime_attr.disk_gb, default_attr.disk_gb]) + " HDD"
+    bootDiskSizeGb: select_first([runtime_attr.boot_disk_gb, default_attr.boot_disk_gb])
+    preemptible: select_first([runtime_attr.preemptible_tries, default_attr.preemptible_tries])
+    maxRetries: select_first([runtime_attr.max_retries, default_attr.max_retries])
+  }
 }
 
-task addRG {
+task addReadGroupAndSort {
     
-    input {
-        File bam_downsample_file
-        String downsample_docker
-        Int disk_size
-        String mem_size
-        File original_cram_file
-    }
-    
-    String bam_readgroup_name = basename(bam_downsample_file, "_downsample.bam") + "_downsample_rg.bam"
-    String bam_sorted_name = basename(bam_downsample_file, "_downsample.bam") + "_downsample_rg_sorted.bam"
+  input {
+    File bam_downsample_file
+    String downsample_docker
+    File original_cram_or_bam_file_read_groups
+    RuntimeAttr? runtime_attr_override
+  }
+  
+  Int num_cpu = 1
+  Int mem_size_gb = 3
+  Int vm_disk_size = 70
+
+  RuntimeAttr default_attr = object {
+    cpu_cores: num_cpu,
+    mem_gb: mem_size_gb, 
+    disk_gb: vm_disk_size,
+    boot_disk_gb: 10,
+    preemptible_tries: 3,
+    max_retries: 1
+  }  
+
+  RuntimeAttr runtime_attr = select_first([runtime_attr_override, default_attr])
+
+  String bam_readgroup_name = basename(bam_downsample_file, "_downsample.bam") + "_downsample_rg.bam"
+  String bam_sorted_name = basename(bam_downsample_file, "_downsample.bam") + "_downsample_rg_sorted.bam"
 
     
-    output {
-        File bam_sorted_rg_file = bam_sorted_name
-    }
+  output {
+    File bam_sorted_rg_file = bam_sorted_name
+  }
 
-    command <<<
-       
-        #re-add read group
-        SM=$(samtools view -H ~{original_cram_file} | grep '^@RG' | head -1 | cut -f7 | sed 's/SM://g')
+  command <<<
+    set -euo pipefail
+    
+    #Re-add read group, necessary for collectCountCrams. This is not best practice. This only adds one RG line, when there are multiples
```

Comment:
I'm not sure on this, I will ask Harrison and others on their input. If I remember correctly, we were trying to do best practice with GATK or picard but it would require some work arounds because of input requirements..